### PR TITLE
2022 02 08 C++ updates

### PR DESCRIPTION
**Pull request by zultron**
This PR is based on #12 and should be reviewed after that PR's resolution.

This PR fixes some problems with the original C++ port; major changes:
- Fix segfault on shutdown, caused by not stopping the Node, Context etc. properly
- Add a `reset` HAL pin meant to reset controllers (see below)
- Work around a bug in Machinekit HAL where too many log messages hang the HAL component load RPC

Minor changes:
- Move C preprocessor macro definitions to `CMakeLists.txt` for streamlining
- Add controller manager startup debug statements to help determine the location of issues

-----
In re. controller reset:  The ROS1 `ros_control` controller manager `update()` method had a `reset_controllers` argument that, when True, would reset controller command to current feedback.  This was needed for uses cases such as after initial power on, where the controller initially commands zero but a battery-backed encoder reports non-zero feedback, enabling the drives causes an unexpected jump.

In ROS2, command reset functionality is moved out to the various controllers' `on_activate()` method.  It's useful for this to be triggered directly from the hardware interface, not only after power on, but also at other times, such as after homing a joint.  This PR adds a `reset` boolean HAL pin.  When the pin goes high, a periodic check in a (non-RT) thread will arrange for controllers to be deactivated & reactivated (at the next RT update, within one cycle).

### CODE REVIEW

**Comment by vatanaksoytezer** [comment]
@zultron can you rebase this onto main and make it ready for review, if you are ready?

**Comment by zultron** [comment]
Thanks for the review on #12, @vatanaksoytezer!  This PR is rebased, and tests are all green.  Ready for review.

File: hal_hw_interface/include/hal_hw_interface/hal_system_interface.hpp [hunk 1]
```diff
@@ -32,8 +32,11 @@
 #ifndef HAL_HW_INTERFACE__HAL_SYSTEM_INTERFACE_HPP_
 #define HAL_HW_INTERFACE__HAL_SYSTEM_INTERFACE_HPP_
 
+#ifndef CNAME
+#error "CNAME must be defined"
```
Comment (vatanaksoytezer) [inline hal_hw_interface/include/hal_hw_interface/hal_system_interface.hpp:36]: Is this possible, I think you already have a cname given by default?

**Review by vatanaksoytezer — DISMISSED** [review]
Just one small question below, rest looks good to me

File: hal_hw_interface/include/hal_hw_interface/hal_system_interface.hpp (hunk 1 above)
Comment (zultron) [reply hal_hw_interface/include/hal_hw_interface/hal_system_interface.hpp:36]: The idea behind that was the `CNAME` macro, defined in the top-level `CMakeLists.txt`, carries the HAL component name to the separate control node source & system interface source, where it must be identical on both sides.  This was intended to help stop me down the road from doing something dumb (which I always do).  You're right, though, it's really not necessary.  If I remove the macro from `CMakeLists.txt`, the compiler barfs up lots of obvious warnings about what's gone wrong.  I'll remove it.

Thanks!

**Review by vatanaksoytezer — APPROVED** [review]
lgtm

**Pull request merged**